fdbclient/MultiVersionTransaction.actor.cpp

```diff
@@ -2052,98 +2093,100 @@ ThreadFuture<Void> MultiVersionDatabase::DatabaseState::monitorProtocolVersion()
 			TraceEvent("ErrorGettingClusterProtocolVersion")
 			    .error(cv.getError())
 			    .detail("ExpectedProtocolVersion", expected);
-		}
 
-		ProtocolVersion clusterVersion =
-		    !cv.isError() ? cv.get() : self->dbProtocolVersion.orDefault(currentProtocolVersion());
-		onMainThreadVoid([self, clusterVersion]() { self->protocolVersionChanged(clusterVersion); });
+			if (self->initializationState == InitializationState::INITIALIZING) {
+				// A failure to retrieve the protocol error is a fatal error, such as invalid or
+				// missing cluster file. There is no point of retrying on it.
+				// Mark the database as failed and abort pending futures
+				self->initializationError = cv.getError();
+				self->initializationState = InitializationState::INITIALIZATION_FAILED;
+				self->dbVar->set(Reference<IDatabase>(), true);
+			}
+		} else {
+			clusterVersion = cv.get();
+			onMainThreadVoid([self, clusterVersion]() { self->protocolVersionChanged(clusterVersion); });
+		}
 		return ErrorOr<Void>(Void());
 	});
 }
 
 // Called when a change to the protocol version of the cluster has been detected.
 // Must be called from the main thread
 void MultiVersionDatabase::DatabaseState::protocolVersionChanged(ProtocolVersion protocolVersion) {
-	if (closed) {
+	if (initializationState == InitializationState::CLOSED) {
 		return;
 	}
 
-	// If the protocol version changed but is still compatible, update our local version but keep the same connection
+	// If the protocol version changed but is still compatible, update our local version but keep the
+	// same connection
 	if (dbProtocolVersion.present() &&
 	    protocolVersion.normalizedVersion() == dbProtocolVersion.get().normalizedVersion()) {
 		dbProtocolVersion = protocolVersion;
-
-		ASSERT(protocolVersionMonitor.isValid());
-		protocolVersionMonitor.cancel();
```

Comment:
What caused this change?

Reply:
I thought it is redundant, because protocolVersionChanged is scheduled at completion of monitorProtocolVersion. My understanding the cancellation makes sense only if the future could be possibly still in the waiting state, but maybe I am overlooking something.

Reply:
I guess the only question is whether this works correctly from the in the legacy case where `protocolVersionChanged` is called from `runGrvProbe`. It might be possible that this could work without canceling the existing one, but in that case we wouldn't want to start a new one. Canceling and replacing should also continue to work, though.